### PR TITLE
#39: Fixed bug related to consuming all offsets given empty records

### LINKED ISSUE #39: Kafka record consumer fails to recognize when it's *already* at the last offset
If I run an import with CONSUME_ALL_OFFSETS = 'true' on a topic where I'm already caught up, the polling loop will keep spinning until a new message appears on the topic, as the `lastRecordOffset` is initialized to 0 instead of the current `partitionEndOffset` -- and then, if no records appeared during the poll, 0 would be *compared* to the `partitionEndOffset` in `shouldContinue()`:

https://github.com/exasol/kafka-connector-extension/blob/7fc77ad6b3f96cbae4993b52b5f386e30f7b76af/src/main/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumer.scala#L70-L95

The fix would probably be to just initialize `lastRecordOffset` to the value of `partitionEndOffset` before the polling loop.

However, not sure what happens if the topic partition is actually empty? Would the Kafka consumer return the end offset as 0? In this case `partitionEndOffset` would become -1, `lastRecordOffset` would also be initialized to -1, and all is fine.

### PR DESCRIPTION
**Pull request by morazow**
Fixes #39 

In addition started refactoring integration tests for #9.

### CODE REVIEW

**Comment by morazow** [comment]
Hello @jwarlander,

Could you please have a look?

I could not test the behavior using integration tests, but instead using mocking in unit tests. It is not very easy to simulate the polling in integration tests.

One other issue is, once consumer obtains empty records it stop and does not continue, even last offset is not reached. I think this should be fine.

**Comment by jwarlander** [comment]
> One other issue is, once consumer obtains empty records it stop and does not continue, even last offset is not reached. I think this should be fine.

Empty records should not.. stop, but rather emit a row full of null values, right? And then proceed. Otherwise it might be a bit troublesome, as we are definitely seeing empty records a *lot* throughout the day in some of our topics :sunglasses: 

Anyhow, I'll take a look at the changes! :smile:

**Comment by morazow** [comment]
Yes, I think it is wrong to call them empty records. Basically, the poll returns nothing, e.g, empty consumer records. Do you have such cases?

Please have a look to [these tests](https://github.com/exasol/kafka-connector-extension/pull/40/files#diff-4737a7ec2240078eadfc1e9fff6509ee5cf8bc350b05ffd5e067da4e33a34bdfR66-R82).

We could also track the record offsets outside of the function, that should allow us to continue until last offset.

**Comment by jwarlander** [comment]
Yeah, ok, I see what you mean now :)

But yes, one of the reasons that I felt an option for full load until latest offset was needed, is that it was very tricky to get a good set of parameters for the Kafka consumer polling. The default parameters apparently ended up with us often getting empty record sets from the polling, so the consumer would stop reading even though there were millions of records left to catch up on.

Interestingly, often I saw that on the first attempt, we'd get say 6-7 million records (out of about 15 million).. then on the second run, just 2-3 million, then maybe 1 million, and so on.. so I don't know if some kind of throttling was happening on our end.

Anyhow, yes, to be clear -- we really need to keep polling even if a single poll returns an empty set :smile:

**Comment by morazow** [comment]
Hey @jwarlander,

Thanks for the feedback! 

Yes, we could definitely continue when getting empty records in between. This assumes, that we would at some point reach the end offset. 

However, maybe we can add another counter, that is if we get some consecutive number of empty polls stop the import.

**Comment by jwarlander** [comment]
Right, or a timeout? Eg. if we haven't gotten *any* records for X seconds (probably 30 by default), stop the import. I feel it's easier / more straightforward to control it this way, than having to think about polling timeout multiplied by max empty poll retries..  Then you can think about those two parameters separately, instead of having to also adjust the other every time you change one of them.

So the cases that are interesting here, I'd say, and when we should stop the import:
- Topic is empty (meaning endOffsets(...) returns 0)
- We're already caught up (partitionStartOffset = partitionEndOffset)
- We've just received empty sets from polling for X seconds (maybe due to factors outside of our control)

Another case I thought of, that's more of a true "edge case" I guess, is -- what if we import from a topic that's been abandoned, and while the offset says X, all the records have actually been purged from the topic already due to age? Here, we'd potentially start out at offset 0 (if it's the first import), or perhaps X-10.. but never get any records, so we'd never "catch up" to the max offset.

This one would be the only instance I know of that we could keep getting 0 records back no matter how many times we poll, so it would be I guess a good reason to have a total "import timeout" as above.

**Comment by morazow** [comment]
> We've just received empty sets from polling for X seconds (maybe due to factors outside of our control)

We already have this, consumer [client waits for 30s](https://github.com/exasol/kafka-connector-extension/blob/97020157ee342173685ed3fb5c8dc3d7abb1da97/src/main/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumer.scala#L49) (by default) until it gets any records from a partition.

In general, I think for now we should assume that we reach the last offset of a partition since we query that end offset beforehand.

I am going add changes so that it continues given in between empty records.

**Comment by jwarlander** [comment]
I pushed a few changes + tests that I think should cover what we need :smile:

**Review by jwarlander — APPROVED** [review]
Looks good as far as I can tell; I added a small suggestion for the changelog!

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+trait BaseDockerIntegrationTest extends AnyFunSuite with BeforeAndAfterAll {
+  private[this] val JAR_DIRECTORY_PATTERN = "scala-"
+  private[this] val JAR_NAME_PATTERN = "exasol-kafka-connector-extension-"
+  private[this] val DEFAULT_EXASOL_DOCKER_IMAGE = "7.0.10"
+
+  val network = DockerNamedNetwork("kafka-it-tests", true)
+  val exasolContainer = {
+    val c: ExasolContainer[_] = new ExasolContainer(getExasolDockerImageVersion())
+    c.withExposedPorts(8563, 2580)
```
Comment (AnastasiiaSergienko) [inline src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:23]: I think these ports ae exposed by default

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  var schema: ExasolSchema = _
+  val assembledJarName = getAssembledJarName()
+
+  override def beforeAll(): Unit =
+    exasolContainer.start()
+
+  override def afterAll(): Unit =
+    exasolContainer.stop()
+
+  def prepareExasolDatabase(schemaName: String): Unit = {
```
Comment (AnastasiiaSergienko) [inline src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:38]: Maybe rename to "installKafkaConnector" or something similar?

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    println(s"Executing: $sql")
+    getConnection().createStatement().execute(sql)
+    ()
+  }
+
+  def executeQuery(sql: String): java.sql.ResultSet =
+    getConnection().createStatement().executeQuery(sql)
+
+  private[this] def getConnection(): java.sql.Connection =
+    exasolContainer.createConnection("")
```
Comment (AnastasiiaSergienko) [inline src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:56]: Are we closing the connections?

File: src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      .assert(
+        table()
+          .row("one", 0L, 0L)
+          .row("two", 0L, 1L)
+          .row("three", 0L, 2L)
+          .matches(NO_JAVA_TYPE_CHECK)
+      )
+  }
+
+  def updateExasolHostsFile(): Unit = {
```
Comment (AnastasiiaSergienko) [inline src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala:75]: Should this also go to Base class?

File: src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  test("emits all records using min and max record counts with empty records") {
+    when(consumer.poll(defaultTimeout))
+      .thenReturn(recordBatch(Seq(0, 1)), emptyConsumerRecords, recordBatch(Seq(2, 3)))
+    KafkaImportChecker().assertEmitCount(2)
+  }
+
+  test("emits all records using consume_all_offsets") {
+    when(consumer.poll(defaultTimeout))
+      .thenReturn(recordBatch(Seq(0, 1)), recordBatch(Seq(2, 3)))
+    KafkaImportChecker(Map("CONSUME_ALL_OFFSETS" -> "true")).assertEmitCount(4)
```
Comment (AnastasiiaSergienko) [inline src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala:69]: What would happen if we set CONSUME_ALL_OFFSETS and MAX_RECORDS_PER_RUN at the same time?

Comment (AnastasiiaSergienko) [reply src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala:69]: Should we add a test case for it?

**Review by AnastasiiaSergienko — COMMENTED** [review]
I left a few small comments :)

File: src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala (hunk 5 above)
Comment (jwarlander) [reply src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala:69]: Feels like that should probably fail and give an explanatory error message back to the user, right? 🤔

Comment (morazow) [reply src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala:69]: Consuming all offset will override other two parameters, I think it should be already described on the user guide. I do not think we need to validate them for now.

Comment (morazow) [reply src/test/scala/com/exasol/cloudetl/kafka/consumer/KafkaRecordConsumerTest.scala:69]: Added test to cover this case

File: src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala (hunk 4 above)
Comment (morazow) [reply src/it/scala/com/exasol/cloudetl/kafka/docker/KafkaImportIT.scala:75]: I was not sure about this method yet, I think I am going to refactor it in #9.

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala (hunk 3 above)
Comment (morazow) [reply src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:56]: Done.

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala (hunk 2 above)
Comment (morazow) [reply src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:38]: Done.

File: src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala (hunk 1 above)
Comment (morazow) [reply src/it/scala/com/exasol/cloudetl/kafka/docker/BaseDockerIntegrationTest.scala:23]: Fixed

**Review by AnastasiiaSergienko — APPROVED** [review]
Looks good! :)

**Pull request merged**